proposals/002-mdm-profiles-sync.md

```diff
@@ -0,0 +1,58 @@
+# MDM Configuration Profiles
+
+## Summary
+
+Fleet MDM will need to store and sync configuration profiles to MDM enrolled devices.
+Configuration profiles are sometimes referred to as “MacOS settings” in the Fleet UI.
+Whenever there is a change to configuration profiles or a device enrolls, configuration profiles will need to be synced to the device.
+
+## Storage
+
+- Store the profiles in a `mdm_apple_configuration_profiles` table with the following columns (incomplete)
+    - `payload_identifier`: this should be unique, but stay the same when editing an existing profile. Use something like `com.fleetdm.<random uuid>`.
+    Built-in profiles that are included with fleet should use a hard coded identifier so that we can update these profiles.
+    This may be required when new options are added or bug fixes are required to an existing profile
+    - `payload_uuid`: this should change whenever an existing profile is updated
+        - note, this doesn't actually need to be a uuid
+    - `version`: isn't part of the actually configuration profile. Used to determine what the latest version of a profile is. For simplicity, this can be an integer that is incremented from 1.
```

Comment:
is the implicit idea here that we never do `UPDATE mdm_apple_configuration_profiles SET raw = ?`, but `INSERT` a new version instead? If yes I like it, but maybe we could use `created_at` instead of tracking a `version`?

Reply:
I thought about using `created_at`, but was worried it might be a little too clever. In theory, anything that could be easily compared would work. The advantage of using `created_at` though would be that we don't need to individually increment a counter per `payload_uuid`.

Reply:
> we don't need to individually track and increment a counter per payload_uuid.

gotcha, thanks! and yes that's the only thing that motivated my comment, but I can see arguments each way.

Reply:
IIRC MySQL timestamps are truncated at seconds. It is a pretty edge case, but do you have any concerns about uniqueness if we are relying on MySQL timestamps?

Reply:
If the `mdm_apple_configuration_profiles` table is insert-only, would we be able to give each row in the table an `id` rely on an auto-incrementing/indexing that instead for easier comparisons?

Reply:
> A DATETIME or TIMESTAMP value can include a trailing fractional seconds part in up to microseconds (6 digits) precision. In particular, any fractional part in a value inserted into a DATETIME or TIMESTAMP column is stored rather than discarded.
> https://dev.mysql.com/doc/refman/5.7/en/datetime.html

seems like we should be fine on that front, but yeah, let's choose whatever is easier to understand later and simple to compare

Reply:
@roperzh, thanks! Somewhat related topic, but perhaps we should consider altering some of the column definitions for other tables to store fractional seconds, particularly where we're doing more time-sensitive async operations that could run on different instances.

> To define a column that includes a fractional seconds part, use the syntax type_name(fsp), where type_name is [TIME](https://dev.mysql.com/doc/refman/5.7/en/time.html), [DATETIME](https://dev.mysql.com/doc/refman/5.7/en/datetime.html), or [TIMESTAMP](https://dev.mysql.com/doc/refman/5.7/en/datetime.html), and fsp is the fractional seconds precision. For example:
> 
> `CREATE TABLE t1 (t TIME(3), dt DATETIME(6));`

Reply:
An auto increment id column could be used as well to determine the latest version.